arch/arm/compare256_neon.c

```diff
@@ -0,0 +1,61 @@
+/* compare256_neon.c - NEON version of compare256
+ * Copyright (C) 2022 Nathan Moinvaziri
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#if defined(ARM_NEON) && defined(HAVE_BUILTIN_CTZLL)
+#ifdef _M_ARM64
+#  include <arm64_neon.h>
+#else
+#  include <arm_neon.h>
+#endif
+#include "../../zbuild.h"
+#include "../../fallback_builtins.h"
+
+static inline uint32_t compare256_neon_static(const uint8_t *src0, const uint8_t *src1) {
+    uint32_t len = 0;
+
+    do {
+        uint8x16_t a, b, cmp;
```

Comment:
We want to try to align one of the two loads here.  In real world use cases, it was definitely a benefit to x86 hardware that cared about alignment.  See: https://github.com/zlib-ng/zlib-ng/blob/develop/arch/x86/compare256_sse2.c#L36

For the most part, the added complexity usually doesn't hurt the cases where you'd be matching in the first vector and where you'd match in the last vector (where you'd have to do 2+ extra compares) seems highly unlikely for most real world use cases (even so, the extra XORs are likely to be inexpensive).

Reply:
Agreed. I think it might be better as a separate PR. That way we can benchmark those changes separately. I doubt I'd able to tell the performance difference in qemu.

Reply:
With aligning one of the inputs, it might not unroll the whole loop anymore as length is no longer guaranteed to be multiple of 16... I tried changing to half-width Neon and it didn't unroll anymore using latest gcc trunk.

There is slight penalty when moving 64 bit elements from vector registers as on ARM it is translated as move from floating-point register to integer register.

Reply:
Is it any faster to move 4 32 bit values, though? It seems like having more jumps and compares would probably eat into those gains.